### PR TITLE
feat: add support for QueryOptions

### PR DESCRIPTION
**Pull request by olavloite**
Adds the ability to set `QueryOptions` when running Cloud Spanner queries through JDBC. For now, only setting the `query_optimizer_version` is added.

QueryOptions can be configured through the following mechanisms:

1. In the connection URL of a JDBC connection using the property `optimizerVersion`. The version specified here will be applied to all queries executed by the connection.
1. By executing the statement `SET OPTIMIZER_VERSION='<version>'` on a JDBC connection. All queries that are executed after this statement will use the version specified.
1. Through the `SPANNER_OPTIMIZER_VERSION` environment variable.

If the options are configured through multiple mechanisms then:

1. Options set at the JDBC connection URL level will override options configured at the environment variable level.
1. Options set by executing a `SET OPTIMIZER_VERSION='<version>'` will override options set at a JDBC connection URL level or environment variable level.

If no options are set, the optimizer version will default to:

1. The optimizer version the database is pinned to.
1. If the database is not pinned to a specific version, then the Cloud Spanner backend will use the "latest" version.

Depends on changes in the client library that have not yet been merged, meaning that the build of this PR will fail until the required changes have been merged.

### CODE REVIEW

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java [hunk 1]
```diff
@@ -1114,4 +1189,74 @@ public void testAddRemoveTransactionRetryListener() {
...
+      impl.executeQuery(Statement.of("SELECT FOO FROM BAR"));
+      verify(unitOfWork)
+          .executeQuery(
+              StatementParser.INSTANCE.parse(
+                  Statement.newBuilder("SELECT FOO FROM BAR")
+                      .withQueryOptions(QueryOptions.newBuilder().setOptimizerVersion("1").build())
+                      .build()),
+              AnalyzeMode.NONE);
+
+      // Execute query with an optimizer version and statistics package set on the connection.
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]:
```suggestion
      // Execute query with an optimizer version set on the connection.
```

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java [hunk 2]
```diff
@@ -1114,4 +1189,74 @@ public void testAddRemoveTransactionRetryListener() {
...
+      impl.executeQuery(Statement.of("SELECT FOO FROM BAR"));
+      verify(unitOfWork)
+          .executeQuery(
+              StatementParser.INSTANCE.parse(
+                  Statement.newBuilder("SELECT FOO FROM BAR")
+                      .withQueryOptions(QueryOptions.newBuilder().setOptimizerVersion("2").build())
+                      .build()),
+              AnalyzeMode.NONE);
+
+      // Execute query with an optimizer version and statistics package set on the connection and
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]:
```suggestion
      // Execute query with an optimizer version set on the connection and
```

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java [hunk 3]
```diff
@@ -1114,4 +1189,74 @@ public void testAddRemoveTransactionRetryListener() {
...
+      verify(unitOfWork)
+          .executeQuery(
+              StatementParser.INSTANCE.parse(
+                  Statement.newBuilder("SELECT FOO FROM BAR")
+                      .withQueryOptions(QueryOptions.newBuilder().setOptimizerVersion("3").build())
+                      .build()),
+              AnalyzeMode.NONE,
+              prefetchOption);
+
+      // Execute query with an optimizer version and statistics package set on the connection, and
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]:
```suggestion
      // Execute query with an optimizer version set on the connection, and
```

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          new SpannerOptions.SpannerEnvironment() {
+            @Override
+            public String getOptimizerVersion() {
+              return "20";
+            }
+
+            @Override
+            public String getOptimizerStatisticsPackage() {
+              return "auto_20200127_20_24_19UTC";
+            }
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java]:
```suggestion

```

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                        QueryOptions.newBuilder()
+                            .setOptimizerVersion("user-defined-version")
+                            .build())
+                    .build())) {
+          assertThat(rs.next()).isTrue();
+          assertThat(rs.getLong(0)).isEqualTo(COUNT_BEFORE_INSERT);
+          assertThat(rs.next()).isFalse();
+
+          ExecuteSqlRequest request = getLastExecuteSqlRequest();
+          // Optimizer version should come from the connection.
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java]:
```suggestion
          // Optimizer version should come from the query.
```

File: src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcQueryOptionsTest.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+/**
+ * {@link QueryOptions} can be used with the JDBC driver on three different levels:
+ *
+ * <ol>
+ *   <li>Specify {@link QueryOptions} in environment variables (TODO: Set environment variables
+ *       runtime)
+ *   <li>Specify {@link QueryOptions} in the connection URL or by setting these using <code>
+ *       SET &lt;query_option&gt; = '&lt;value&gt;'</code> statements.
+ *   <li>Specify {@link QueryOptions} in a query hint (TODO: Check SQL syntax)
```
Comment (skuruppu) [inline src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcQueryOptionsTest.java]: Could these TODOs be resolved now or do they need to be done later?

**Comment by olavloite** [comment]
> LGTM with some minor comments.
> 
> One request, if you could update the description of this PR to contain the information about the env variable and the precedence, that would be great. Example [PR](https://github.com/googleapis/google-cloud-dotnet/pull/4512)

Done.

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java (hunk 1 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]: Done

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java (hunk 2 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]: Done.

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java (hunk 3 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/ConnectionImplTest.java]: Done.

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java (hunk 4 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java]: Done

File: src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcQueryOptionsTest.java (hunk 6 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcQueryOptionsTest.java]: The first one was already done (although it depends on changes in the client library that have not yet been merged).

I've checked the syntax for query hints and added a test for that as well, and removed the TODO.

File: src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java (hunk 5 above)
Comment (olavloite) [reply src/test/java/com/google/cloud/spanner/jdbc/ConnectionTest.java]: Done.

**Comment by skuruppu** [comment]
I merged in the [v1.51.0](https://github.com/googleapis/java-spanner/pull/96) release which should hopefully be available on Maven by the end of your day.

**Comment by skuruppu** [comment]
Looks like I can't merge in #75 until #73 can be merged in but that can only be done once googleapis/java-core#179 is in a release. Yikes, so many dependencies!

**Comment by skuruppu** [comment]
I don't understand this failure in the [Linkage Monitor](https://source.cloud.google.com/results/invocations/1214d3a9-1878-41f3-b624-8827ae45181f/targets/cloud-devrel%2Fclient-libraries%2Fjava%2Fjava-spanner-jdbc%2Fpresubmit%2Flinkage-monitor/log) when all other tests pass:
```
Mar 17, 2020 2:24:16 AM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor run
INFO: BOM Coordinates: com.google.cloud:libraries-bom:4.2.0
Mar 17, 2020 2:27:17 AM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor run
SEVERE: Newly introduced problems:
Class com.google.spanner.v1.ExecuteSqlRequest$QueryOptions is not found
  referenced from com.google.cloud.spanner.jdbc.ConnectionImpl (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
  referenced from com.google.cloud.spanner.jdbc.ConnectionOptions (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
  referenced from com.google.cloud.spanner.jdbc.StatementParser (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
Class com.google.spanner.v1.ExecuteSqlRequest$QueryOptions$Builder is not found
  referenced from com.google.cloud.spanner.jdbc.ConnectionImpl (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
  referenced from com.google.cloud.spanner.jdbc.ConnectionOptions (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
  referenced from com.google.cloud.spanner.jdbc.StatementParser (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
(google-cloud-spanner-1.49.2.jar) com.google.cloud.spanner.Statement's method getQueryOptions() is not found
  referenced from com.google.cloud.spanner.jdbc.StatementParser (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
(google-cloud-spanner-1.49.2.jar) com.google.cloud.spanner.Statement's method toBuilder() is not found
  referenced from com.google.cloud.spanner.jdbc.StatementParser (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)
(google-cloud-spanner-1.49.2.jar) com.google.cloud.spanner.Statement$Builder's method withQueryOptions(com.google.spanner.v1.ExecuteSqlRequest$QueryOptions arg1) is not found
  referenced from com.google.cloud.spanner.jdbc.StatementParser (google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar)

google-cloud-spanner-jdbc-1.13.1-SNAPSHOT.jar is at:
  com.google.cloud:google-cloud-spanner-jdbc:1.13.1-SNAPSHOT (compile)
google-cloud-spanner-1.49.2.jar is at:
  com.google.cloud:google-cloud-spanner:1.49.2 (compile)

Mar 17, 2020 2:27:17 AM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor main
SEVERE: Found 5 new linkage errors
```

**Comment by hengfengli** [comment]
@skuruppu I keep receiving notification emails due to the label `automerge`. I just removed it at the moment.

**Comment by olavloite** [comment]
I don't understand this Linkage Monitor error either. I just tried to execute it locally and it executed without any problems. I also don't understand why the log seems to indicate that Cloud Spanner 1.49.2 is being used, while it clearly uses 1.51.0. The output locally was:

```
loite@loite-XPS-15-9575:~/wsgoogle/java-spanner-jdbc$ mv ~/Downloads/linkage-monitor-latest-all-deps.jar .
loite@loite-XPS-15-9575:~/wsgoogle/java-spanner-jdbc$ java -jar linkage-monitor-latest-all-deps.jar com.google.cloud:libraries-bom
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Mar 17, 2020 2:09:30 PM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor run
INFO: BOM Coordinates: com.google.cloud:libraries-bom:4.2.0
Mar 17, 2020 2:12:29 PM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor run
INFO: Snapshot versions have the same 816 errors as baseline
Mar 17, 2020 2:12:29 PM com.google.cloud.tools.dependencies.linkagemonitor.LinkageMonitor main
INFO: No new problem found
```

**Comment by olavloite** [comment]
@skuruppu The IT failure has been fixed, but the error in the Linkage Monitor step is preventing this from being merged.

**Comment by skuruppu** [comment]
@chingor13 would you be able to help us with the two unexplained test failures.
1. We don't know why the Linkage Monitor is failing. It seems like it's relying on an older version of the client library despite having updated the pom.xml file.
2. The samples tests will not pass since this change needs to be merged in before it can pass.

Any help is appreciated.

**Comment by suztomo** [comment]
(Linkage Monitor developer here). The failure looks good to me; no need to fix it.

GCP Libraries BOM 4.2.0 has com.google.cloud:google-cloud-spanner:1.49.2
https://storage.googleapis.com/cloud-opensource-java-dashboard/com.google.cloud/libraries-bom/4.2.0/artifact_details.html. The error means if we upgrade google-cloud-spanner-jdbc's version in the BOM, there would be a conflict. This means we need to upgrade google-cloud-spanner-jdbc and google-cloud-spanner together next time we update the content of the BOM.

> We don't know why the Linkage Monitor is failing. It seems like it's relying on an older version of the client library despite having updated the pom.xml file.

GCP Library BOM does not look at google-cloud-spanner-jdbc's pom.xml.

**Comment by skuruppu** [comment]
> The error means if we (GCP Libraries BOM maintainer) upgrade only google-cloud-spanner-jdbc's version in the BOM, there would be a conflict with google-cloud-spanner:1.49.2. Therefore we (GCP Libraries BOM maintainer) need to upgrade google-cloud-spanner-jdbc and google-cloud-spanner together next time we update the content of the BOM.

Thanks for the explanation @suztomo. When will the BOM contents be updated? This PR needs to be merged in before the end of the week as it's a blocker for a feature launch. We can't merge it currently because the PR requires the Linkage Monitor step to pass.

**Comment by suztomo** [comment]
I don't think of a better way than skipping the check.

- libraries-bom:4.2.0 has `google-cloud-spanner-bom:1.49.2` through [`com.google.cloud:google-cloud-bom:0.122.3-alpha`](https://search.maven.org/artifact/com.google.cloud/google-cloud-bom/0.122.3-alpha/pom)
- The latest [`com.google.cloud:google-cloud-bom:0.122.4-alpha`](https://search.maven.org/artifact/com.google.cloud/google-cloud-bom/0.122.4-alpha/pom) has `google-cloud-spanner-bom:1.50.0` (not the latest 1.51.0).
- Therefore, if we wait for libraries-bom to have `google-cloud-spanner:1.51.0`, the BOM needs to wait for a new google-cloud-bom that contains `google-cloud-spanner:1.51.0`.

@elharo Do you think of a better way than either skipping the check or waiting for google-cloud-bom release?

**Comment by elharo** [comment]
The reason this check is here is that having versions out of sync like this causes hard-to-debug problems for customers. Looking at the output of the linkage monitor, this is exactly the sort of thing they're going to hit.

I think @suztomo's analysis is correct. google-cloud-bom should push a new version that includes google-cloud-spanner:1.51.0. You don't need to wait for this to happen. Just ask @chingor13 to release it. Once that's out--it takes about 4 hours from start to finish, mostly waiting on Sonatype to publish--we can push a new version of libraries-bom that pulls that in.

**Comment by skuruppu** [comment]
Thanks @elharo and @suztomo for the explanations. @chingor13 would you be able to help us with the google-cloud-bom release?

**Comment by chingor13** [comment]
Yep, I can do one

**Comment by skuruppu** [comment]
Looks like [java-cloud-bom/releases/tag/v0.123.2](https://github.com/googleapis/java-cloud-bom/releases/tag/v0.123.2) has [java-cloud-bom/pull/406](https://github.com/googleapis/java-cloud-bom/pull/406). Are there any other steps required @elharo?

**Comment by elharo** [comment]
we still need a libraries-bom release that incorporates this. One is in progress.

**Comment by suztomo** [comment]
@chingor13 Can you rerun the Linkage Monitor check? [GCP Libraries BOM 4.3.0](https://search.maven.org/artifact/com.google.cloud/libraries-bom/4.3.0/pom) has been released today.

**Comment by suztomo** [comment]
@chingor13 I see it succeeded. Thank you.

**Comment by skuruppu** [comment]
Thanks so much @chingor13, @suztomo and @elharo for your help with this. We really appreciate it.

**Pull request merged**